sway/desktop/render.c

```diff
@@ -150,6 +164,185 @@ static void render_texture(struct wlr_output *wlr_output,
+	}
+}
+
+/** Blurs the main_buffer content and returns the blurred framebuffer */
+struct fx_framebuffer *get_main_buffer_blur(struct fx_renderer *renderer,
+		struct sway_output *output, pixman_region32_t *original_damage,
+		const float _matrix[static 9], const float box_matrix[static 9],
+		const struct wlr_box *box) {
+	int width, height;
+	wlr_output_transformed_resolution(output->wlr_output, &width, &height);
+
+	/* Blur */
+	const enum wl_output_transform transform = wlr_output_transform_invert(output->wlr_output->transform);
+	float matrix[9];
+	struct wlr_box monitor_box = { 0, 0, width, height };
+	wlr_matrix_project_box(matrix, &monitor_box, transform, 0, output->wlr_output->transform_matrix);
+
+	float gl_matrix[9];
+	wlr_matrix_multiply(gl_matrix, renderer->projection, matrix);
+
+	pixman_region32_t damage;
+	pixman_region32_init(&damage);
+	pixman_region32_copy(&damage, original_damage);
+	wlr_region_transform(&damage, &damage, wlr_output_transform_invert(output->wlr_output->transform),
+			width, height);
+	wlr_region_expand(&damage, &damage, get_blur_size());
+
+	// Initially blur main_buffer content into the effects_buffers
+	struct fx_framebuffer *current_buffer = &renderer->main_buffer;
+
+	// Bind to blur framebuffer
+	fx_framebuffer_bind(&renderer->effects_buffer, width, height);
+	glBindTexture(renderer->main_buffer.texture.target, renderer->main_buffer.texture.id);
+
+	// damage region will be scaled, make a temp
+	pixman_region32_t tempDamage;
+	pixman_region32_init(&tempDamage);
+	// When DOWNscaling, we make the region twice as small because it's the TARGET
+	wlr_region_scale(&tempDamage, &damage, 0.5f);
+
+	int blur_radius = config->blur_params.radius;
+	int blur_passes = config->blur_params.num_passes;
+
+	// First pass
+	render_blur_segments(renderer, output, gl_matrix, &tempDamage, &current_buffer,
+			&renderer->shaders.blur1, box, blur_radius);
+
+	// Downscale
+	for (int i = 1; i < blur_passes; ++i) {
+		wlr_region_scale(&tempDamage, &damage, 1.0f / (1 << (i + 1)));
+		render_blur_segments(renderer, output, gl_matrix, &tempDamage, &current_buffer,
+				&renderer->shaders.blur1, box, blur_radius);
+	}
+
+	// Upscale
+	for (int i = blur_passes - 1; i >= 0; --i) {
+		// when upsampling we make the region twice as big
+		wlr_region_scale(&tempDamage, &damage, 1.0f / (1 << i));
+		render_blur_segments(renderer, output, gl_matrix, &tempDamage, &current_buffer,
+				&renderer->shaders.blur2, box, blur_radius);
+	}
+
+	pixman_region32_fini(&tempDamage);
+	pixman_region32_fini(&damage);
+
+	// Bind back to the default buffer
+	fx_framebuffer_bind(&renderer->main_buffer, width, height);
+
+	return current_buffer;
+}
+
+void render_blur(bool optimized, struct sway_output *output,
+		pixman_region32_t *output_damage, const struct wlr_fbox *src_box,
+		const struct wlr_box *dst_box, pixman_region32_t *opaque_region,
+		int surface_width, int surface_height, int32_t surface_scale,
```

Comment:
Seems like we only ever pass `1` as the `surface_scale`. Should we remove this?

Reply:
Even in the surface_iterator?

Reply:
Updated to pass the scale